engine/api/v2_repository_analyze.go

```diff
@@ -443,6 +434,91 @@ func (api *API) analyzeRepository(ctx context.Context, projectRepoID string, ana
 	return sdk.WithStack(tx.Commit())
 }
 
+func findCommitter(ctx context.Context, cache cache.Store, db *gorp.DbMap, analysis sdk.ProjectRepositoryAnalysis, vcsProjectWithSecret sdk.VCSProject, repoName string, vcsPublicKeys map[string][]GPGKey) (*sdk.AuthentifiedUser, string, string, error) {
+	publicKeyFound := false
+	publicKeys, has := vcsPublicKeys[vcsProjectWithSecret.Type]
+	if has {
+		for _, k := range publicKeys {
+			if analysis.Data.SignKeyID == k.ID {
+				publicKeyFound = true
+				break
+			}
+		}
+	}
+
+	if !publicKeyFound {
+		gpgKey, err := user.LoadGPGKeyByKeyID(ctx, db, analysis.Data.SignKeyID)
+		if err != nil {
+			if !sdk.ErrorIs(err, sdk.ErrNotFound) {
+				return nil, "", "", sdk.NewErrorFrom(err, "unable get gpg key: %s", analysis.Data.SignKeyID)
+			}
+			return nil, sdk.RepositoryAnalysisStatusSkipped, fmt.Sprintf("gpgkey %s not found", analysis.Data.SignKeyID), nil
+		}
+
+		cdsUser, err := user.LoadByID(ctx, db, gpgKey.AuthentifiedUserID)
+		if err != nil {
+			if !sdk.ErrorIs(err, sdk.ErrNotFound) {
+				return nil, "", "", sdk.WithStack(sdk.NewErrorFrom(err, "unable to load user %s", gpgKey.AuthentifiedUserID))
+			}
+			return nil, sdk.RepositoryAnalysisStatusError, fmt.Sprintf("user %s not found for gpg key %s", gpgKey.AuthentifiedUserID, gpgKey.KeyID), nil
+		}
+		return cdsUser, "", "", nil
+
+	}
+
+	// Get commit
+	tx, err := db.Begin()
+	if err != nil {
+		return nil, "", "", sdk.WithStack(err)
+	}
+	defer tx.Rollback() // nolint
+
+	client, err := repositoriesmanager.AuthorizedClient(ctx, tx, cache, analysis.ProjectKey, vcsProjectWithSecret.Name)
+	if err != nil {
+		return nil, "", "", sdk.WithStack(err)
+	}
+
+	switch vcsProjectWithSecret.Type {
+	case sdk.VCSTypeBitbucketServer, sdk.VCSTypeGitlab:
+		commit, err := client.Commit(ctx, repoName, analysis.Commit)
+		if err != nil {
+			return nil, "", "", err
+		}
+		commitUser, err := user.LoadByUsername(ctx, tx, commit.Committer.Slug)
+		if err != nil {
+			if !sdk.ErrorIs(err, sdk.ErrNotFound) {
+				return nil, "", "", sdk.WithStack(sdk.NewErrorFrom(err, "unable to get user %s", commit.Committer.Slug))
+			}
+			return nil, sdk.RepositoryAnalysisStatusSkipped, fmt.Sprintf("committer %s not found in CDS", commit.Committer.Slug), nil
+		}
+		return commitUser, "", "", nil
+	case sdk.VCSTypeGithub:
+		pr, err := client.SearchPullRequest(ctx, repoName, analysis.Commit, "closed")
+		if err != nil {
+			return nil, "", "", sdk.WithStack(sdk.NewErrorFrom(err, "unable to retrieve pull request with commit %s", analysis.Commit))
+		}
+		// FIXME:  need to map vcs user <-> cds user
```

Comment:
needs work?

Reply:
it will be in the next PR. It needs refact of authen driver and the PR would be too long, so I split in 2